### PR TITLE
Convert a page of historical location forecast data into a flattened numpy array for easier use

### LINKED ISSUE #88: Improve historical data retrieval
### What's needed?

- Instead of returning the protobuf objects, we should return objects that enable custom conversions
- Forecast to numpy array (still same 3 dims)
- Address creation ts filtering on numpy array in client -> Issue:
https://confluence.ecmwf.int/display/DAC/Dissemination+schedule
- Validity_time should be range

### Proposed solution

_No response_

### Use cases

_No response_

### Alternatives and workarounds

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by david-natingga-frequenz**
- Enable the conversion of protobuf LocationForecast list to a flattened numpy
  array.
- Enhance the historical data retrieval example with conversion to
  a pandas dataframe, sample output:
```
(sdk_venv22) david@david-desktop:~/projects/frequenz/source/frequenz-api-weather$ PYTHONPATH=py python examples/iterate_hist_forecast.py "localhost:50051"
            creation_ts latitude longitude         valid_at_ts                                     feature      value
0   2022-06-27 12:00:00     52.5      13.4 2022-06-27 13:00:00  ForecastFeature.U_WIND_COMPONENT_100_METRE  -2.629823
1   2022-06-27 12:00:00     52.5      13.4 2022-06-27 13:00:00  ForecastFeature.V_WIND_COMPONENT_100_METRE   2.316034
2   2022-06-27 12:00:00     52.5      13.4 2022-06-27 14:00:00  ForecastFeature.U_WIND_COMPONENT_100_METRE  -2.439865
3   2022-06-27 12:00:00     52.5      13.4 2022-06-27 14:00:00  ForecastFeature.V_WIND_COMPONENT_100_METRE   1.965858
4   2022-06-27 12:00:00     52.5      13.4 2022-06-27 15:00:00  ForecastFeature.U_WIND_COMPONENT_100_METRE    -1.6119
..                  ...      ...       ...                 ...                                         ...        ...
337 2023-12-18 00:00:00     52.5      13.4 2023-12-18 22:00:00  ForecastFeature.V_WIND_COMPONENT_100_METRE   4.606584
338 2023-12-18 00:00:00     52.5      13.4 2023-12-18 23:00:00  ForecastFeature.U_WIND_COMPONENT_100_METRE  11.596823
339 2023-12-18 00:00:00     52.5      13.4 2023-12-18 23:00:00  ForecastFeature.V_WIND_COMPONENT_100_METRE   4.560944
340 2023-12-18 00:00:00     52.5      13.4 2023-12-19 00:00:00  ForecastFeature.U_WIND_COMPONENT_100_METRE  11.380841
341 2023-12-18 00:00:00     52.5      13.4 2023-12-19 00:00:00  ForecastFeature.V_WIND_COMPONENT_100_METRE   4.240217

[342 rows x 6 columns]
```

Closes https://github.com/frequenz-floss/frequenz-api-weather/issues/88.

### CODE REVIEW

File: examples/iterate_hist_forecast.py [hunk 1]
```diff
@@ -53,11 +54,32 @@ async def main(service_address: str) -> None:
...
+    async for forecasts in location_forecast_iterator:
+        # You can work directly with the protobuf object forecasts.
+        # Here we choose to flatten into into a numpy array instead.
+        arr = forecasts.flatten()
+        arrays.append(arr)
+    data = np.concatenate(arrays)
+
+    # Optionally, you can construct a pandas dataframe from the data.
+    # pylint: disable=import-outside-toplevel, import-error
+    import pandas as pd  # type: ignore[import]
```
Comment (david-natingga-frequenz) [inline examples/iterate_hist_forecast.py:65]: We have a policy not to use pandas in production for now. But this is just an example and we may use it in our experiments.

Comment (david-natingga-frequenz) [reply examples/iterate_hist_forecast.py:65]: Also pylint ignores could not be inline since flake would complain about the line being too long.

File: examples/iterate_hist_forecast.py [hunk 2]
```diff
@@ -53,11 +54,32 @@ async def main(service_address: str) -> None:
...
 
-    async for location_forecast in location_forecast_iterator:
-        for forecasts in location_forecast.forecasts:
-            print("Timestamp:", forecasts.valid_at_ts)
-            for feature_forecast in forecasts.features:
-                print(feature_forecast)
+    arrays = []
+    async for forecasts in location_forecast_iterator:
+        # You can work directly with the protobuf object forecasts.
+        # Here we choose to flatten into into a numpy array instead.
```
Comment (TalweSingh) [inline examples/iterate_hist_forecast.py:60]: remove one into

Comment (TalweSingh) [inline examples/iterate_hist_forecast.py]: remove one into

File: examples/iterate_hist_forecast.py (hunk 1 above)
Comment (TalweSingh) [reply examples/iterate_hist_forecast.py:65]: classic, it's quite annoying

File: py/frequenz/client/weather/_historical_forecast_iterator.py [hunk 3]
```diff
@@ -65,29 +65,26 @@ async def __anext__(self) -> weather_pb2.LocationForecast:
         Raises:
             StopAsyncIteration: If there are no more historical weather forecasts.
         """
-        if len(self.location_forecasts) == 0 and self.page_token == EMPTY_PAGE_TOKEN:
+        if self.page_token == EMPTY_PAGE_TOKEN:
             raise StopAsyncIteration
 
-        if self.location_forecasts is None or len(self.location_forecasts) == 0:
```
Comment (TalweSingh) [inline py/frequenz/client/weather/_historical_forecast_iterator.py:71]: Why did you remove this check?

File: py/frequenz/client/weather/_types.py [hunk 4]
```diff
@@ -176,108 +183,204 @@ def to_ndarray_vlf(
...
+
+        # get the feature indexes of the proto for the filtered features
+        if features:
+            for req_feature in features:
+                for f_index, feature in enumerate(
+                    location_forecasts[0].forecasts[0].features
+                ):
+                    if req_feature == ForecastFeature.from_pb(feature.feature):
+                        feature_indexes.append(f_index)
+                        break
```
Comment (TalweSingh) [inline py/frequenz/client/weather/_types.py]: Looks like somewhat similar code. Maybe it's possible to make this less redundant?
"""# pylint: disable=too-many-locals,too-many-branches,too-many-statements"""
is also a sign of this. But if you think that's fine i will trust your judgement.

File: py/frequenz/client/weather/_types.py [hunk 5]
```diff
@@ -176,108 +183,204 @@ def to_ndarray_vlf(
...
+            )
 
-        # catch all exceptions
-        except Exception as e:
-            raise RuntimeError("Error processing forecast data") from e
+    # catch all exceptions
+    except Exception as e:
+        raise RuntimeError("Error processing forecast data") from e
+
+    return array
```
Comment (TalweSingh) [inline py/frequenz/client/weather/_types.py]: function is quite long in general. Maybe you could split it up

**Review by TalweSingh — COMMENTED** [review]
Looks good in general. But maybe someone else wants to take a look as well. I haven't really worked on the client yet

**Comment by TalweSingh** [comment]
Also, split the code changes to the client and the example in 2 commits. And a more descriptive title for the PR itself would be great

File: py/frequenz/client/weather/_historical_forecast_iterator.py (hunk 3 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py:71]: The reason is that the logic changed here. Before the iterator was over each LocationForecast, now the iterator is over each page that contains `repeated LocationForecast`.

File: py/frequenz/client/weather/_types.py (hunk 4 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py]: A good point. This is not new code. In this PR I just moved it out of Forecasts class. There are actually more problems with it, so I created a separate issue: https://github.com/frequenz-floss/frequenz-api-weather/issues/93

File: py/frequenz/client/weather/_types.py (hunk 5 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py]: I agree. In this PR I just moved the code, I created a separate issue for it: https://github.com/frequenz-floss/frequenz-api-weather/issues/93

**Comment by david-natingga-frequenz** [comment]
> Also, split the code changes to the client and the example in 2 commits. And a more descriptive title for the PR itself would be great

I made the title more descriptive.

As for putting an example into a separate commit, it is problematic because the old example would not work with the new code. The reason is that the historical data iterator is changed in this PR, it returns a different object.

File: examples/iterate_hist_forecast.py (hunk 2 above)
Comment (david-natingga-frequenz) [reply examples/iterate_hist_forecast.py]: Thanks for spotting this.

**Comment by shsms** [comment]
Looks like the lack of tests is starting to be a problem.  I hope you've tested there's no changes in behaviour for live streaming from the changes to shared methods.

**Comment by david-natingga-frequenz** [comment]
> Looks like the lack of tests is starting to be a problem. I hope you've tested there's no changes in behaviour for live streaming from the changes to shared methods.

I think the method to_ndarray_vlf is quite complicated and I do not see in it much use. In https://github.com/frequenz-floss/frequenz-api-weather/issues/93 I propose just to remove it. @shsms Do you know any code depending on that method? In this PR, I just move the method out, but I do not modify it, so its behaviour should not change.

As for my testing, I just run the live streaming locally and it works with this code just as it worked with the code before. However, you are right, there are no tests that would warrant the exact behaviour as before.

**Comment by cwasicki** [comment]
@david-natingga-frequenz This is very difficult to review. One reason is, that there are too many changes in a single commit. Can this be split into separate commits?

> Do you know any code depending on that method?

Yes, our forecasts.

> However, you are right, there are no tests that would warrant the exact behaviour as before.

We should have these.

**Comment by david-natingga-frequenz** [comment]
Based on the feedback of all of you, I simplified this PR and put it into 2 commits.
- I kept the function to_ndarray_vlf and the class _types.Forecasts unchaged. Any work on to_ndarray_vlf can be addressed in another PR (in order to keep this PR small), an issue already created https://github.com/frequenz-floss/frequenz-api-weather/issues/93.
- Instead I created a new class _types.HistoricalForecasts with a function flatten just for the historical data.

@cwasicki Please, have a look.

File: py/frequenz/client/weather/_types.py [hunk 6]
```diff
@@ -281,3 +281,85 @@ def to_ndarray_vlf(
             raise RuntimeError("Error processing forecast data") from e
 
         return array
+
+
+@dataclass(frozen=True)
+class HistoricalForecasts:
+    """Historical weather forecast data."""
+
+    _forecasts_pb: weather_pb2.GetHistoricalWeatherForecastResponse
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py:297]: You could use one of the suggestions [here](https://github.com/frequenz-floss/frequenz-api-common/discussions/215) and hope that it will collect most votes.

File: py/frequenz/client/weather/_types.py [hunk 7]
```diff
@@ -281,3 +281,85 @@ def to_ndarray_vlf(
...
+            ValueError: If the forecasts data is missing or invalid.
+        """
+        # check for empty forecasts data
+        if not self._forecasts_pb.location_forecasts:
+            raise ValueError("Forecast data is missing or invalid.")
+
+        return flatten(list(self._forecasts_pb.location_forecasts))
+
+
+# pylint: disable=too-many-locals,too-many-branches,too-many-statements
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: Is this obsolete?

File: py/frequenz/client/weather/_types.py [hunk 8]
```diff
@@ -281,3 +281,85 @@ def to_ndarray_vlf(
...
+def flatten(
+    location_forecasts: list[weather_pb2.LocationForecast],
+) -> np.ndarray[
+    tuple[typing.Any, typing.Any, typing.Any, typing.Any, typing.Any, typing.Any],
+    np.dtype[np.float64],
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py]: Why is your return type not a 2D array but an array of tuples?

File: py/frequenz/client/weather/_types.py [hunk 9]
```diff
@@ -281,3 +281,85 @@ def to_ndarray_vlf(
...
+            Numpy array of tuples with the flattened forecast data.
+
+        Raises:
+            ValueError: If the forecasts data is missing or invalid.
+        """
+        # check for empty forecasts data
+        if not self._forecasts_pb.location_forecasts:
+            raise ValueError("Forecast data is missing or invalid.")
+
+        return flatten(list(self._forecasts_pb.location_forecasts))
```
Comment (cwasicki) [inline py/frequenz/client/weather/_types.py:328]: Why is this `list` necessary?

File: py/frequenz/client/weather/_historical_forecast_iterator.py [hunk 10]
```diff
@@ -65,29 +65,26 @@ async def __anext__(self) -> weather_pb2.LocationForecast:
...
-                    weather_pb2.GetHistoricalWeatherForecastRequest(
-                        locations=(location.to_pb() for location in self.locations),
-                        features=(feature.value for feature in self.features),
-                        start_ts=self.start_ts,
-                        end_ts=self.end_ts,
-                        pagination_params=pagination_params,
-                    )
-                )
+        pagination_params = pagination_params_pb2.PaginationParams()
+        pagination_params.page_size = PAGE_SIZE
```
Comment (cwasicki) [inline py/frequenz/client/weather/_historical_forecast_iterator.py]: Why is page size a constant and not configurable? Is it optimized?

File: py/frequenz/client/weather/_historical_forecast_iterator.py [hunk 11]
```diff
@@ -65,29 +65,26 @@ async def __anext__(self) -> weather_pb2.LocationForecast:
...
+        if self.page_token is not None:
+            pagination_params.page_token = self.page_token
+
+        response: Any = await self._stub.GetHistoricalWeatherForecast(  # type:ignore
```
Comment (cwasicki) [inline py/frequenz/client/weather/_historical_forecast_iterator.py]: Why `Any`?

File: py/frequenz/client/weather/_historical_forecast_iterator.py [hunk 12]
```diff
@@ -65,29 +65,26 @@ async def __anext__(self) -> weather_pb2.LocationForecast:
...
+                features=(feature.value for feature in self.features),
+                start_ts=self.start_ts,
+                end_ts=self.end_ts,
+                pagination_params=pagination_params,
             )
+        )
 
-            self.page_token = response.pagination_info.next_page_token
-            self.location_forecasts = response.location_forecasts
+        self.page_token = response.pagination_info.next_page_token
```
Comment (cwasicki) [inline py/frequenz/client/weather/_historical_forecast_iterator.py:98]: Are we sure that this will never be `None` after this line?

File: py/frequenz/client/weather/_types.py (hunk 6 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py:297]: Thank you for raising the awareness of the ongoing issue. Since the issue is not resolved, I suggest to keep it this way as it is. It is consistent with Forecasts class and classes for other protobuf objects (e.g. LocationForecast) that already exist here. Once the issue is resolved and conventions change, we can change both Forecasts and HistoricalForecasts classes. I created an issue for this: https://github.com/frequenz-floss/frequenz-api-weather/issues/95

File: py/frequenz/client/weather/_types.py (hunk 8 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py]: It seemed a more natural choice to use tuples for me since I did not want to mix objects of different types within the same numpy array. ChatGPT seems to agree as well (https://chat.openai.com/share/4394f8f0-d423-4f92-8504-5a3d14a872b7):

Choosing between using a 2D NumPy array or a NumPy array of tuples depends on your specific use case and the operations you plan to perform. Here are some considerations to help you decide:

**2D NumPy Array**
- Use Case: Ideal for mathematical operations, linear algebra, and where you need to perform operations on entire rows or columns simultaneously. If your data represents a matrix or a grid (like in image processing or numerical simulations), a 2D array is usually the way to go.
- Performance: Generally offers better performance for vectorized operations, as NumPy is optimized for operations on arrays of uniform data types. This means faster computations and less memory usage compared to a NumPy array of tuples.
- Indexing: Allows for more intuitive and flexible indexing and slicing operations, which can be very convenient when working with subarrays or when you need to extract specific rows, columns, or elements.

**NumPy Array of Tuples**
- Use Case: Suitable when each element of the array needs to hold a fixed number of elements of potentially different types (e.g., a combination of integers, floats, and strings) and these elements together represent a single entity or record. This structure can be useful for tabular data where each tuple represents a row with different fields.
- Flexibility: Provides the flexibility to store elements of different types in each tuple, mimicking structured arrays or records. This can be handy when working with complex data structures that do not fit well into a purely numerical matrix format.
- Operations: Operations on arrays of tuples can be less efficient and more cumbersome than on 2D arrays, especially for mathematical computations. You might find yourself iterating over the array more often, which could lead to slower performance compared to vectorized operations on 2D arrays.

**Decision Factors**
- Data Homogeneity: If your data is homogeneous (all numerical or all of the same type) and you need to perform mathematical computations, go with a 2D NumPy array.
- Data Structure: If your data is structured (like records in a table, each with different types), and you don't need to perform vectorized numerical computations, an array of tuples could be more suitable.
- Performance Needs: For high-performance computing needs, especially with large datasets where you'll benefit from vectorization, a 2D array is typically preferred.

In summary, if your primary concern is performance and ease of mathematical computations, a 2D NumPy array is likely the better choice. If you're dealing with complex data structures that require elements of different types to be grouped together, an array of tuples might be more appropriate, albeit with some performance trade-offs.

File: py/frequenz/client/weather/_types.py (hunk 9 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py:328]: Because the function `flatten` takes as an argument list rather than repeated. I thought it is a better design decision that way as lists are simpler and more standard than repeated objects.

File: py/frequenz/client/weather/_types.py (hunk 7 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py]: A good catch. I am deleting this.

File: py/frequenz/client/weather/_historical_forecast_iterator.py (hunk 10 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: I agree this should be part of the configuration. In PR I kept it as a constant to avoid the additional complexity. There is no configuration for the client now, so we would need to introduce the whole class/module for this rather than just add some value to an existing file.

So I suggest this should be attended to in another PR. I created an issue for this: https://github.com/frequenz-floss/frequenz-api-weather/issues/96

Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: @shsms suggested that it may be hard for a user to deal with the configuration. So how about I create config.py and move the constant there? `config.py` would be part of the code, so a user would not need to deal with that. At the same time it would enable us to gather the configuration and constants in one place.

Comment (shsms) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: I think he just meant take it as a parameter in the method.  Configurable doesn't mean from a file.

Comment (shsms) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: Probably in the `__init__`, and you could give it a default value.

File: py/frequenz/client/weather/_historical_forecast_iterator.py (hunk 11 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: Changing this to `weather_pb2.GetHistoricalWeatherForecastResponse`, thanks.

File: py/frequenz/client/weather/_historical_forecast_iterator.py (hunk 12 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py:98]: The API spec does not guarantee this. This assumption is based on the current service implementation which is subject to a change.

Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py:98]: So I added an additional condition checking for `None` to be able to catch this in case the service implementation changes.

**Comment by david-natingga-frequenz** [comment]
@cwasicki Ready for a review again.

File: py/frequenz/client/weather/_historical_forecast_iterator.py (hunk 10 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: Ok, taking your idea with setting the value in `__init__` and using a default value.

File: py/frequenz/client/weather/_types.py (hunk 8 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py]: TODO: consider using a named tuple.

File: py/frequenz/client/weather/_historical_forecast_iterator.py (hunk 10 above)
Comment (cwasicki) [reply py/frequenz/client/weather/_historical_forecast_iterator.py]: >So I suggest this should be attended to in another PR. I created an issue for this: https://github.com/frequenz-floss/frequenz-api-weather/issues/96

Yes, I mean an parameter in the `__init__` method, sorry if that wasn't clear enough. I don't think we need that issue.

File: py/frequenz/client/weather/_types.py (hunk 8 above)
Comment (david-natingga-frequenz) [reply py/frequenz/client/weather/_types.py]: @cwasicki Replaced numpy arrays with a list of named tuples. Ready for a review again.

**Review by cwasicki — APPROVED** [review]
